src/systems/filecoin_vm/interpreter/vm_interpreter.go

```diff
@@ -123,51 +169,48 @@ func (vmi *VMInterpreter_I) ApplyMessage(inTree st.StateTree, message msg.Unsign
 		),
 	)
 
-	if !sendRet.ExitCode().AllowsStateUpdate() {
-		// error -- revert all state changes -- ie drop updates. burn used gas.
-		outTree = inTree
-		outTree = _withTransferFundsAssert(
-			outTree,
-			message.From(),
-			addr.BurntFundsActorAddr,
-			_gasToFIL(sendRet.GasUsed(), message.GasPrice()),
-		)
-	} else {
-		// success -- refund unused gas
-		outTree = sendRetStateTree
-		refundGas := message.GasLimit() - sendRet.GasUsed()
-		TODO() // TODO: assert refundGas is nonnegative
-		outTree = _withTransferFundsAssert(
-			outTree,
-			addr.BurntFundsActorAddr,
-			message.From(),
-			_gasToFIL(refundGas, message.GasPrice()),
-		)
+	ok = _vmiUseGas(sendRet.GasUsed())
+	if !ok {
+		panic("Interpreter error: runtime execution used more gas than provided")
 	}
 
-	outTree, err = outTree.Impl().WithIncrementedCallSeqNum(message.To())
-	if err != nil {
-		// TODO: if actor deletion is possible at some point, may need to allow this case
-		panic("Internal interpreter error: failed to increment call sequence number")
+	ok = _vmiUseGas(gascost.OnChainReturnValue(sendRet.ReturnValue()))
+	if !ok {
+		return _applyError(compTreePreSend, exitcode.OutOfGas, addr.BurntFundsActorAddr, minerAddr)
+	}
+
+	compTreeRet := compTreePreSend
+	if sendRet.ExitCode().AllowsStateUpdate() {
+		compTreeRet = compTreePostSend
 	}
 
-	// reward miner gas fees
-	outTree = _withTransferFundsAssert(
-		outTree,
+	// Refund unused gas to sender.
+	refundGas := vmiGasRemaining
+	compTreeRet = _withTransferFundsAssert(
+		compTreeRet,
 		addr.BurntFundsActorAddr,
-		minerAddr, // TODO: may not exist
-		_gasToFIL(sendRet.GasUsed(), message.GasPrice()),
+		message.From(),
+		_gasToFIL(refundGas, message.GasPrice()),
 	)
 
-	return outTree, sendRet
+	return _applyReturn(
+		compTreeRet, vmr.InvocOutput_Make(sendRet.ReturnValue()), sendRet.ExitCode(),
+		addr.BurntFundsActorAddr, minerAddr)
 }
 
-func _applyError(errCode exitcode.SystemErrorCode) vmr.MessageReceipt {
-	// TODO: should this gasUsed value be zero?
-	// If nonzero, there is not guaranteed to be a nonzero gas balance from which to deduct it.
-	gasUsed := gascost.ApplyMessageFail
-	TODO()
-	return vmr.MessageReceipt_MakeSystemError(errCode, gasUsed)
+func _stateTreeWithGasUsedFundsTransfer(
+	tree st.StateTree, gasUsed msg.GasAmount, gasPrice msg.GasPrice,
+	fundsFrom addr.Address, fundsTo addr.Address) st.StateTree {
+
+	TODO() // TODO: what if the miner has insufficient funds?
+	// Should we instead aggregate these deltas, and SubtractWhileNonnegative at the end of the tipset?
```

Comment:
Blocking: please resolve this in this PR.

Note that the miner (owner) will just have been paid the block reward before processing any other messages, so can be penalised at least up to that amount. But this code should still define a state transition that will succeed.